### ISSUE TITLE
Simplify Tween, remove some stuff, make it easier to maintain.

### ISSUE THREAD

**Issue by trusktr** [tweenjs/tween.js#655]
Proposed (welcome to debate):

- [ ] Remove the `TWEEN` singleton.
  - Teach people how to make a Tween and update it directly.
  - Teach them that if they want to update multiple Tweens at once, they can make a Group.
  - This eliminates possible errors people will have when they clash tweens from two or more separate pieces of code.
- [ ] Remove relative values. It involves detecting strings for little convenience. People can just use math on their side with not much more effort.
- [ ] Remove the dynamic-to feature. It isn't used often, and much internal complication.
- [ ] Remove array of to-values.
  - Also adds quite some complexity.
  - It seems that when we do this, we inadvertently destroy the meaning of an easing curve because the actual number output is no longer on that curve so the meaning is obscure, unless we do very special math to determine proper values to use such that a new curve makes sense.
  - The same math skill required for the previous point may instead be used to chain multiple tweens together, which gives more capabilities such as specific curves between each value. To emulate the array of values, chain tweens together with linear curves, then interpolate all of them with a single higher-level tween using a single curve.
  - The previous point allows much more options than the inflexible feature of to-values. To-values is a shortcut for the previous point with a linear chain and directing tween.
  - Downside: lose ability to choose interpolation method, f.e. bezier, which is useful for making a smooth curve out of the discrete to-values.
    - provide an alternative that doesn't complicate tween directly? F.e. Three.js has Path and Curve tools, instead animate position along such primitives specialized for that purpose. Perhaps we show a demo of that here, after removing to-values.

What else?

**Comment by trusktr** [comment]
@MasatoMakino hello! :) If you have some time, we could jam on these ideas.

**Comment by MasatoMakino** [comment]
@trusktr hello!

#### `TWEEN` singleton.

> docs: Teach them that if they want to update multiple Tweens at once, they must make a Group.

Does this mean the [mainGroup](https://github.com/tweenjs/tween.js/blob/e01bbdbebe2d05509b73bbbe7dc7438228124a9e/src/Tween.ts#L52C62-L52C62) is deprecated?

#### Remove the dynamic-to and array of to-values feature

I agree with the removal of dynamic-to and to-array. These are very large features. If necessary, I think they should be separated into independent npm packages or replaced with reliable packages.

I agree with the policy of duplicating tween in utility functions rather than adding functionality to tween instances. it makes `Tween.ts` easier to maintain.

#### Deprecated steps

Due to the removal of feature, Tween.js users should be given time to prepare.

- Add [deprecated](https://typedoc.org/tags/deprecated/) comment to tsdoc
- Display a warning in the browser console if a deprecated feature is used.
- Change the major version of the package after removing the feature.

To display warning messages, you can refer to deprecation.ts in pixi.js. This approach is user-friendly because the same warning is not displayed more than once.

https://github.com/pixijs/pixijs/blob/v7.2.4/packages/utils/src/logging/deprecation.ts

#### Visual test

This is related to what is core of Tween.js.

1. Tween.js is a library for value completion.
1. Tween.js is a library for drawing animations.

If 1, then visual testing is not necessary. Tween.js is only responsible for outputting reproducible values, not for rendering.
If it is 2, then browser testing is required.

I use Tween.js in combination with three.js and Pixi.js. In this case, Tween.js is responsible for 1.

**Comment by trusktr** [comment]
> Does this mean the [mainGroup](https://github.com/tweenjs/tween.js/blob/e01bbdbebe2d05509b73bbbe7dc7438228124a9e/src/Tween.ts#L52C62-L52C62) is deprecated?

Yeah, that would be removed. Instead, people would:

- create and control individual tweens
- create groups if they want to control multiple tweens

Similar to the good practice of avoiding global variables.

---

Pretty much agree with all you said. We definitely need to release the breaking changes in a major version.

Tween has numerical outputs, but a few visual tests might catch some edge cases. Not sure if worth it yet.

**Comment by nya13** [comment]
dynamic-to option is an amazing thing and should not be removed.  I've used it and that adds AI-like behaviour to the tween.

The Group object still needs a lot of work.  Pausing tweens inside a Group terminates them completely, which is not correct.  We should be able to do a .startAll() and .pauseAll() on a Group.  A .restartAll() and .reverseAll() would be great also.  Of course, those functions used on a single tween would be good also.

**Comment by trusktr** [comment]
@nya13 Thanks for the input!

> Pausing tweens inside a Group terminates them completely, which is not correct

That may indeed be true. The current`TWEEN` group should have the same behavior, as it is just a `Group`, so I'd consider it a separate issue. Can you please open a separate ticket for that?

> dynamic-to option is an amazing thing and should not be removed. I've used it and that adds AI-like behaviour to the tween.

Indeed I agree the *use case* can be desirable, but perhaps the way we currently achieve that use case breaks the concept of a Tween (the curve on which a number transitions between two values), plus makes the Tween class a bunch more complicated.

Instead, I believe we can come up with cleaner ways to achieve the same effects. For example, if we keep Tween simple, then we can instead animate along a path, and modify the path dynamically (f.e. modify the path connecting 🐇 that the 🦊 is chasing). In that rabbit-fox example, the concept is actually two concepts that are currently blended together in an unclear way: we have a tween, *and* we're animating along a path (and the path happens to change shape dynamically). If we update that example to clearly show those two concepts separately, the example will be semantically clear while the Tween class can be simplified.

Can you provide any more use cases?

**Comment by trusktr** [comment]
I added a new one to the list:

> Remove the `autostart` parameter from `tween.update(time, autostart)`. It makes the API less intuitive: for example if you call `stop()` and then the next animation frame calls `update()` and restarts it, the tween seems like it does not stop. Better to make the APIs explicit. Tweens will start and stop only when you tell them to, regardless if `update` is being called.

**Comment by ronaldcurtis** [comment]
> remove legacy CommonJS

Would be good if we don't remove CommonJS compatibility just yet, as [Electron](https://www.electronjs.org/) does not yet support ES modules [link](https://github.com/electron/electron/issues/21457).

**Comment by trusktr** [comment]
@ronaldcurtis Good point! Haven't removed CJS yet (in fact, the CJS `exports` condition in package.json is fixed).

Looks like ESM support in Electron finally landed in August!

Are you using Tween.js with Electron? If so, can you try it out? If it works, we'll mark the CommonJS module as deprecated with a `console.warn`, and plan to remove it after some time.

**Comment by ronaldcurtis** [comment]
> Looks like ESM support in Electron finally landed in August! Docs:
🎉 

@trusktr Not using Tween.js anymore, but as you mentioned it looks like Electron supports ESM now, so it should be fine to deprecate ^_^

**Comment by trusktr** [comment]
@dalisoft updates the list above to check off items done in 

- #713

**Comment by dalisoft** [comment]
@trusktr I got, we make it happen. Refactor and make it modern

**Comment by trusktr** [comment]
fixed typo in my last comment: "updates" -> "I updated"
